privacyidea/lib/sqlutils.py

```diff
@@ -0,0 +1,118 @@
+# License along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+#
+
+from sqlalchemy import select, text
+from sqlalchemy.ext.compiler import compiles
+from sqlalchemy.sql import ClauseElement, Delete
+
+
+class DeleteLimit(Delete, ClauseElement):
+    """
+    A modified DELETE clause element that allows to specify a limit
+    for the number of rows that should be deleted.
+
+    Deleting a large number of rows via just one SQL statement
+    may cause deadlocks in a replicated setup. This clause can
+    be used to split the one big DELETE statement into multiple
+    smaller statements ("chunks") which reduces the probability
+    of deadlocks.
+
+    A dedicated clause element is needed because different
+    databases use different notation for limits on DELETE statements.
+    """
+    def __init__(self, table, filter, limit=1000):
+        Delete.__init__(self, table)
+        self.table = table
+        self.filter = filter
+        if not isinstance(limit, int):
+            raise RuntimeError('limit must be an integer')
+        if limit <= 0:
+            raise RuntimeError('limit must be positive')
+        self.limit = limit
+
+
+@compiles(DeleteLimit)
+def visit_delete_limit(element, compiler, **kw):
+    """
+    Default compiler for the DeleteLimit clause element.
+    This compiles to a DELETE statement with a SELECT subquery which
+    has a limit set::
+
+        DELETE FROM ... WHERE id IN
+        (SELECT id FROM ... WHERE ... LIMIT ...)
+
+    However, this syntax is not supported by MySQL.
+    """
+    select_stmt = select([element.table.c.id]).where(element.filter).limit(element.limit)
+    delete_stmt = element.table.delete().where(element.table.c.id.in_(select_stmt))
+    return compiler.process(delete_stmt)
+
+
+@compiles(DeleteLimit, 'mysql')
+def visit_delete_limit_mysql(element, compiler, **kw):
+    """
+    Special compiler for the DeleteLimit clause element
+    for MySQL dialects. This compiles to a DELETE element
+    with a LIMIT::
+
+        DELETE FROM pidea_audit WHERE ... LIMIT ...
+    """
+    return 'DELETE FROM {} WHERE {} LIMIT {:d}'.format(
+        compiler.process(element.table, asfrom=True),
+        compiler.process(element.filter), element.limit)
+
+
+def delete_chunked(session, table, filter, limit=1000):
+    """
+    Delete all rows matching a given filter criterion from a table,
+    but only delete *limit* rows at a time. Commit after each DELETE.
+
+    :param session: SQLAlchemy session object
+    :param table: SQLAlchemy table object (e.g. ``LogEntry.__table__``)
+    :param filter: A filter criterion (e.g. ``LogEntry.age < now``)
+    :param limit: Number of rows to delete in one chunk
+    :return: total number of deleted rows
+    """
+    deleted = 0
+    statement = DeleteLimit(table, filter, limit)
+    while True:
+        result = session.execute(statement)
+        deleted += result.rowcount
+        session.commit()
+        if result.rowcount < limit:
```

Comment:
What does "result.rowcount" contain?

I need to delete 1500 entries with a chungsize of 1000.
Would this mean, first 1000 entries would be deleted and then 500? The rowcount would be smaller the ``limit`` -> exit.
What would happen if there are exacty 2000 entries?

There would be a final run with the attempt to delete up to 1000 entries, but no entires available.
Would this do any harm?

Reply:
It is fine. I merge it